packages/smart-contracts/test/contracts/ERC20BatchConversionPayments.test.ts

```diff
@@ -0,0 +1,611 @@
+      conversionsToPay_results = conversionsToPay_results.concat([
+        conversionToPay.result,
+        conversionToPay2.result,
+      ]);
+      conversionsFees_results = conversionsFees_results.concat([
+        conversionFees.result,
+        conversionFees2.result,
+      ]);
+    }
+
+    // get balances of the 2nd token, useful when there are 2 different tokens used
+    const fromOldBalance2 = await testERC20b.balanceOf(from);
+    const toOldBalance2 = await testERC20b.balanceOf(to);
+    const feeOldBalance2 = await testERC20b.balanceOf(feeAddress);
+
+    const tx = await batchConvFunction(argTemplate(convDetails), feeAddress);
+    if (logGas) {
+      const receipt = await tx.wait();
+      console.log(`${2 * nTimes} req, gas consumption: `, receipt.gasUsed.toString());
+    }
+
+    // 1st condition: every tokens (end of the paths) are identicals
+    if (
+      convDetail.path[convDetail.path.length - 1] === convDetail2.path[convDetail2.path.length - 1]
+    ) {
+      [fromDiffBalanceExpected, toDiffBalanceExpected, feeDiffBalanceExpected] =
+        expectedERC20Balances(conversionsToPay_results, conversionsFees_results, batchConvFee);
+    }
+    // else: there are 2 different tokens used (end of the paths): testERC20 and testERC20b
+    else {
+      // calculate the expected balances of the 1st token: testERC20
+      [fromDiffBalanceExpected, toDiffBalanceExpected, feeDiffBalanceExpected] =
+        expectedERC20Balances(
+          conversionsToPay_results.filter((_, i) => i % 2 === 0),
+          conversionsFees_results.filter((_, i) => i % 2 === 0),
+          batchConvFee,
+        );
+
+      // calculate the expected balances of the 2nd token: testERC20b
+      const [fromDiffBalanceExpected2, toDiffBalanceExpected2, feeDiffBalanceExpected2] =
+        expectedERC20Balances(
+          conversionsToPay_results.filter((_, i) => i % 2 === 1),
+          conversionsFees_results.filter((_, i) => i % 2 === 1),
+          batchConvFee,
+        );
+
+      // check the balance of testERC20b token, which is not checked in "afterEach" as testERC20 token.
+      checkBalancesForOneToken(
+        testERC20b,
+        fromOldBalance2,
+        toOldBalance2,
+        feeOldBalance2,
+        fromDiffBalanceExpected2,
+        toDiffBalanceExpected2,
+        feeDiffBalanceExpected2,
+      );
+    }
+  };
+
+  /**
+   * @notice it contains all the tests related to the ERC20 batch conversion payment, and its context required
+   * @param isBatchRouter allows to use the function "batchERC20ConversionPaymentsMultiTokens"
+   *                     through the batchRouter or directly
+   */
+  const ERC20ConversionTestSuite = (isBatchRouter: boolean) => {
+    before(() => {
+      setBatchConvFunction(isBatchRouter, signer);
+    });
+
+    describe(isBatchRouter ? 'Through batchRouter' : 'Without batchRouter', () => {
+      describe('batchERC20ConversionPaymentsMultiTokens with DAI', async () => {
```

Comment:
I think the function definition + calls can be eased with a simple loop through 2 values, and does not require the `before` section

Reply:
Done for the loop.
Regarding before, we need it for every functions, I don't see another way really cleaner